File: app/src/main/java/com/arcinteractive/spaces/data/model/SpacePermission.kt
package com.arcinteractive.spaces.data.model

enum class SpacePermission(
    val wireValue: String,
    val title: String
) {
    ManageSpaceSettings("manage_space_settings", "Manage Space Settings"),
    ManageMembers("manage_members", "Manage Members"),
    ManageRoles("manage_roles", "Manage Roles"),
    ManageModules("manage_modules", "Manage Modules"),
    InviteMembers("invite_members", "Invite Members"),
    RemoveMembers("remove_members", "Remove Members"),
    CreateEvents("create_events", "Create Events"),
    PostPings("post_pings", "Post Pings"),
    UploadFiles("upload_files", "Upload Files"),
    UploadPhotosVideos("upload_photos_videos", "Upload Photos/Videos"),
    EditOwnContent("edit_own_content", "Edit Own Content"),
    DeleteOwnContent("delete_own_content", "Delete Own Content"),
    DeleteOthersContent("delete_others_content", "Delete Others' Content"),
    ViewContent("view_content", "View Content")
}

val SpaceMemberRole.capabilities: Set<SpacePermission>
    get() = when (this) {
        SpaceMemberRole.Owner -> SpacePermission.entries.toSet()
        SpaceMemberRole.Admin -> setOf(
            SpacePermission.ManageMembers,
            SpacePermission.ManageRoles,
            SpacePermission.ManageModules,
            SpacePermission.InviteMembers,
            SpacePermission.RemoveMembers,
            SpacePermission.CreateEvents,
            SpacePermission.PostPings,
            SpacePermission.UploadFiles,
            SpacePermission.UploadPhotosVideos,
            SpacePermission.EditOwnContent,
            SpacePermission.DeleteOwnContent,
            SpacePermission.DeleteOthersContent,
            SpacePermission.ViewContent
        )
        SpaceMemberRole.Moderator -> setOf(
            SpacePermission.CreateEvents,
            SpacePermission.PostPings,
            SpacePermission.UploadFiles,
            SpacePermission.UploadPhotosVideos,
            SpacePermission.EditOwnContent,
            SpacePermission.DeleteOwnContent,
            SpacePermission.DeleteOthersContent,
            SpacePermission.ViewContent
        )
        SpaceMemberRole.Member -> setOf(
            SpacePermission.CreateEvents,
            SpacePermission.PostPings,
            SpacePermission.UploadFiles,
            SpacePermission.UploadPhotosVideos,
            SpacePermission.EditOwnContent,
            SpacePermission.DeleteOwnContent,
            SpacePermission.ViewContent
        )
        SpaceMemberRole.Guest -> setOf(SpacePermission.ViewContent)
    }

fun SpaceMemberRole.canRemove(targetRole: SpaceMemberRole, isTargetOwner: Boolean): Boolean {
    if (isTargetOwner) return false

    return when (this) {
        SpaceMemberRole.Owner -> true
        SpaceMemberRole.Admin -> targetRole != SpaceMemberRole.Admin
        SpaceMemberRole.Moderator, SpaceMemberRole.Member, SpaceMemberRole.Guest -> false
    }
}

fun SpaceMemberRole.canChangeRole(
    targetRole: SpaceMemberRole,
    newRole: SpaceMemberRole,
    isTargetOwner: Boolean
): Boolean {
    if (isTargetOwner || newRole == SpaceMemberRole.Owner) return false

    return when (this) {
        SpaceMemberRole.Owner -> targetRole != SpaceMemberRole.Owner
        SpaceMemberRole.Admin -> {
            val isAllowedTarget = targetRole == SpaceMemberRole.Moderator ||
                targetRole == SpaceMemberRole.Member ||
                targetRole == SpaceMemberRole.Guest
            val isAllowedNewRole = newRole == SpaceMemberRole.Moderator ||
                newRole == SpaceMemberRole.Member ||
                newRole == SpaceMemberRole.Guest
            isAllowedTarget && isAllowedNewRole
        }
        SpaceMemberRole.Moderator, SpaceMemberRole.Member, SpaceMemberRole.Guest -> false
    }
}

fun Space.roleFor(userId: String?, members: List<SpaceMember>): SpaceMemberRole? {
    if (userId == null) return null
    if (userId == ownerId) return SpaceMemberRole.Owner
    return members.firstOrNull { it.id == userId }?.role
}

fun Space.hasCapability(
    userId: String?,
    members: List<SpaceMember>,
    capability: SpacePermission
): Boolean {
    val role = roleFor(userId, members) ?: return false
    return role.capabilities.contains(capability)
}

File: iOS/Spaces V2/Models/SpacePermission.swift
import Foundation

enum SpacePermission: String, CaseIterable, Identifiable, Hashable {
    case manageSpaceSettings = "manage_space_settings"
    case manageMembers = "manage_members"
    case manageRoles = "manage_roles"
    case manageModules = "manage_modules"
    case inviteMembers = "invite_members"
    case removeMembers = "remove_members"
    case postPings = "post_pings"
    case createEvents = "create_events"
    case uploadPhotosVideos = "upload_photos_videos"
    case uploadFiles = "upload_files"
    case editOwnContent = "edit_own_content"
    case deleteOwnContent = "delete_own_content"
    case deleteOthersContent = "delete_others_content"
    case viewContent = "view_content"

    var id: String { rawValue }

    var title: String {
        switch self {
        case .manageSpaceSettings: "Manage Space Settings"
        case .manageMembers: "Manage Members"
        case .manageRoles: "Manage Roles"
        case .manageModules: "Manage Modules"
        case .inviteMembers: "Invite Members"
        case .removeMembers: "Remove Members"
        case .postPings: "Post Pings"
        case .createEvents: "Create Events"
        case .uploadPhotosVideos: "Upload Photos/Videos"
        case .uploadFiles: "Upload Files"
        case .editOwnContent: "Edit Own Content"
        case .deleteOwnContent: "Delete Own Content"
        case .deleteOthersContent: "Delete Others' Content"
        case .viewContent: "View Content"
        }
    }
}

extension SpaceMemberRole {
    var capabilities: Set<SpacePermission> {
        switch self {
        case .owner:
            return Set(SpacePermission.allCases)
        case .admin:
            return [
                .manageMembers,
                .manageRoles,
                .manageModules,
                .inviteMembers,
                .removeMembers,
                .postPings,
                .createEvents,
                .uploadPhotosVideos,
                .uploadFiles,
                .editOwnContent,
                .deleteOwnContent,
                .deleteOthersContent,
                .viewContent
            ]
        case .moderator:
            return [
                .postPings,
                .createEvents,
                .uploadPhotosVideos,
                .uploadFiles,
                .editOwnContent,
                .deleteOwnContent,
                .deleteOthersContent,
                .viewContent
            ]
        case .member:
            return [
                .postPings,
                .createEvents,
                .uploadPhotosVideos,
                .uploadFiles,
                .editOwnContent,
                .deleteOwnContent,
                .viewContent
            ]
        case .guest:
            return [.viewContent]
        }
    }

    func canRemove(targetRole: SpaceMemberRole, isTargetOwner: Bool) -> Bool {
        if isTargetOwner {
            return false
        }

        switch self {
        case .owner:
            return true
        case .admin:
            return targetRole != .admin
        case .moderator, .member, .guest:
            return false
        }
    }

    func canChangeRole(of targetRole: SpaceMemberRole, to newRole: SpaceMemberRole, isTargetOwner: Bool) -> Bool {
        if isTargetOwner || newRole == .owner {
            return false
        }

        switch self {
        case .owner:
            return targetRole != .owner
        case .admin:
            guard targetRole == .moderator || targetRole == .member || targetRole == .guest else {
                return false
            }
            return newRole == .moderator || newRole == .member || newRole == .guest
        case .moderator, .member, .guest:
            return false
        }
    }
}

extension Space {
    func role(for userID: String?, members: [SpaceMember]) -> SpaceMemberRole? {
        guard let userID else { return nil }
        if userID == ownerId {
            return .owner
        }
        return members.first(where: { $0.id == userID })?.role
    }

    func hasCapability(_ capability: SpacePermission, for userID: String?, members: [SpaceMember]) -> Bool {
        guard let role = role(for: userID, members: members) else {
            return false
        }
        return role.capabilities.contains(capability)
    }
}
